src/java/org/apache/cassandra/net/async/RebufferingByteBufDataInputPlus.java

```diff
@@ -0,0 +1,225 @@
+    private final int highWaterMark;
+    private final ChannelConfig channelConfig;
+
+    private volatile boolean closed;
+
+    public RebufferingByteBufDataInputPlus(int lowWaterMark, int highWaterMark, ChannelConfig channelConfig)
+    {
+        super(Unpooled.EMPTY_BUFFER.nioBuffer());
+
+        if (lowWaterMark > highWaterMark)
+            throw new IllegalArgumentException(String.format("low water mark is greater than high water mark: %d vs %d", lowWaterMark, highWaterMark));
+
+        currentBuf = Unpooled.EMPTY_BUFFER;
+        this.lowWaterMark = lowWaterMark;
+        this.highWaterMark = highWaterMark;
+        this.channelConfig = channelConfig;
+        queue = new LinkedBlockingQueue<>();
+        queuedByteCount = new AtomicInteger();
+    }
+
+    /**
+     * Append a {@link ByteBuf} to the end of the einternal queue.
+     *
+     * Note: it's expected this method is invoked on the netty event loop.
+     */
+    public void append(ByteBuf buf) throws IllegalStateException
+    {
+        assert buf != null : "buffer cannot be null";
+
+        if (closed)
+        {
+            ReferenceCountUtil.release(buf);
+            throw new IllegalStateException("stream is already closed, so cannot add another buffer");
+        }
+
+        // this slightly undercounts the live count as it doesn't include the currentBuf's size.
+        // that's ok as the worst we'll do is allow another buffer in and add it to the queue,
+        // and that point we'll disable auto-read. this is a tradeoff versus making some other member field
+        // atomic or volatile.
+        int queuedCount = queuedByteCount.addAndGet(buf.readableBytes());
+        if (channelConfig.isAutoRead() && queuedCount > highWaterMark)
+            channelConfig.setAutoRead(false);
+
+        queue.add(buf);
+    }
+
+    /**
+     * {@inheritDoc}
+     *
+     * Release open buffers and poll the {@link #queue} for more data.
+     * <p>
+     * This is best, and more or less expected, to be invoked on a consuming thread (not the event loop)
+     * becasue if we block on the queue we can't fill it on the event loop (as that's where the buffers are coming from).
+     */
+    @Override
+    protected void reBuffer() throws IOException
+    {
+        currentBuf.release();
+        buffer = null;
+        currentBuf = null;
+
+        // possibly re-enable auto-read, *before* blocking on the queue, because if we block on the queue
+        // without enabling auto-read we'll block forever :(
+        if (!channelConfig.isAutoRead() && queuedByteCount.get() < lowWaterMark)
+            channelConfig.setAutoRead(true);
+
+        try
+        {
+            while (true)
+            {
+                // consume any buffers from the queue *before* checking (and acting on) the closed status.
+                // that allows us to consume any last buffers before the end-of-stream.
+                currentBuf = queue.poll(1, TimeUnit.SECONDS);
```

Comment:
Hmm, i like this better than what I have. I'm not a fan what I had with the `poll(long, TImeUnit)`, but it allowed you out of a blocking scenario. 

wrt to the `while` loop, I think i need to keep that in case there's a spurious wake up from `take()`, in which case it throws an `InterruptedException`. Does this makes sense, or am I mistaken?